taichi/platform/opengl/opengl_api.cpp

```diff
@@ -0,0 +1,270 @@
+       source for GL drawing and image specification commands.
+   ***/
+
+  GLSSBO &bind_data(void *data, size_t size, GLuint usage = GL_STATIC_READ)
+  {
+    glBindBuffer(GL_SHADER_STORAGE_BUFFER, id_);
+    glBufferData(GL_SHADER_STORAGE_BUFFER, size, data, usage);
+    return *this;
+  }
+
+  GLSSBO &bind_index(size_t index)
+  {
+    // SSBO index, is `layout(std430, binding = <index>)` in shader.
+    // We use only one SSBO though...
+    glBindBufferBase(GL_SHADER_STORAGE_BUFFER, index, id_);
+    return *this;
+  }
+
+  void *map(size_t offset, size_t length, GLbitfield access = GL_MAP_READ_BIT)
+  {
+    // map GPU memory to CPU address space, offset within SSBO data
+    return glMapBufferRange(GL_SHADER_STORAGE_BUFFER, offset, length, access);
+  }
+};
+
+void initialize_opengl()
+{
+    glfwInit();
+    // Compute Shader requires OpenGL 4.3+ (or OpenGL ES 3.1+)
+    glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 4);
+    glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
+    // GLEW cannot load GL without a context
+    // And the best way to make context is by creating a window
+    // Then hide it immediately, LOL
+    GLFWwindow *window = glfwCreateWindow(1, 1, "Make GLEW Happy", nullptr, nullptr);
+    if (!window) {
+      const char *desc = nullptr;
+      GLint status = glfwGetError(&desc);
+      if (!desc) desc = "Unknown Error";
+      TI_ERROR("[glsl] cannot create GLFW window: error {}: {}", status, desc);
+    }
+    glfwHideWindow(window);
+    glfwMakeContextCurrent(window);
+#ifdef USE_GLEW
+    GLint status = glewInit();
+    if (status != GLEW_OK) {
+      TI_ERROR("[glsl] cannot initialize GLEW: {}", glewGetErrorString(status));
+    }
+#endif
+    const char *gl_version = (const char *)glGetString(GL_VERSION);
+    if (!gl_version) {
+      TI_WARN("[glsl] cannot get OpenGL version");
+    } else {
+      TI_INFO("[glsl] OpenGL {}", gl_version);
+    }
+}
+
+void *launch_glsl_kernel(std::string source, void *data, size_t data_size)
+{
+  static bool gl_inited = false;
+  if (!gl_inited) {
+    // TODO: move this to somewhere like main()
+    initialize_opengl();
+    gl_inited = true;
+  }
+
+  GLShader shader(source);
+  GLProgram program(shader);
+  program.link();
+  program.use();
+
+  GLSSBO ssbo;
+  ssbo.bind_index(0);
+  ssbo.bind_data(data, data_size, GL_DYNAMIC_READ); // input
+
+  // https://www.khronos.org/opengl/wiki/Compute_Shader
+  // https://community.arm.com/developer/tools-software/graphics/b/blog/posts/get-started-with-compute-shaders
+  // https://www.khronos.org/assets/uploads/developers/library/2014-siggraph-bof/KITE-BOF_Aug14.pdf
+  //
+  // `glDispatchCompute(X, Y, Z)`   - the X*Y*Z  == `Blocks`   in CUDA
+  // `layout(local_size_x = X) in;` - the X      == `Threads`  in CUDA
+  //
+  glDispatchCompute(1, 1, 1);
+  glMemoryBarrier(GL_SHADER_STORAGE_BARRIER_BIT);
+
+  void *data_r = ssbo.map(0, data_size); // output
```

Comment:
Map GPU memory into current address space, at data_r. Then use `set_arg<int>` or `memcpy` to move result back to ctx/global.